src/coreclr/jit/utils.h

```diff
@@ -773,4 +775,138 @@ int64_t GetSigned64Magic(int64_t d, int* shift /*out*/);
 
 double CachedCyclesPerSecond();
 
+namespace CheckedOps
+{
+const bool Unsigned = true;
+const bool Signed   = false;
+
+// Important note: templated functions below must use dynamic "assert"s instead of "static_assert"s
+// because they can be instantiated on code paths that are not reachable at runtime, but visible
+// to the compiler. One example is VN's EvalOp<T> function, which can be instantiated with "size_t"
+// for some operators, and that's legal, but its callee EvalOpSpecialized<T> uses "assert(!AddOverflows(v1, v2))"
+// for VNF_ADD_OVF/UN, and would like to continue doing so without casts.
+
+template <class T>
+bool AddOverflows(T firstAddend, T secondAddend, bool unsignedAdd)
+{
+    typedef typename std::make_unsigned<T>::type UT;
+    assert((std::is_same<T, int32_t>::value || std::is_same<T, int64_t>::value));
+
+    T result = firstAddend + secondAddend;
```

Comment:
Signed integer overflow is "undefined behavior", we should probably avoid using it in new code.

Reply:
Heh, yes, that's one of the things that surprised me when I first started to work with this code. We practically rely on it being defined for correctness in folding (in `gtFoldExprConst` and VN, I am sure in other places too).

So far `-fwrapv` has been protecting us on Unix, but there is no equivalent flag being used for MSVC, so we have just been getting lucky there.

Reply:
If we depend on `-fwrapv` in many other places, then I guess it is OK.